File: Game/AttackCalculator.cs
using System;

namespace Game
{
    public class AttackCalculator
    {
        Random random = new Random();

        public int CalculateDamage(Character atk, Character def)
        {
            var dice = random.Next(1, 20);
            return CalculateDamage(atk, def, dice);
        }

        public int CalculateDamage(Character atk, Character def, int dice)
        {
            var isAttackerAbleToDamage = atk.Force + dice > def.armorClass;
            return isAttackerAbleToDamage ? DamageDealt(atk, dice) : 0;
        }


        private int DamageDealt(Character atk, int dice)
        {
            switch (dice)
            {
                case 1:
                    return 0;
                case 20:
                    return atk.damageDealt * 2;
                default:
                    return atk.damageDealt;
            }
        }
    }
}

File: kotlin/src/main/kotlin/game/AttackCalculator.kt
package game

import kotlin.random.Random
import kotlin.random.nextInt

class AttackCalculator {
    private val random = Random

    fun calculateDamage(atk: Character, def: Character): Int {
        val dice = random.nextInt(1..20)
        return calculateDamage(atk = atk, def = def, dice = dice)
    }

    fun calculateDamage(atk: Character, def: Character, dice: Int): Int {
        val defaultAttack = atk.force
        val currentAttack = defaultAttack + dice
        var damage = atk.damageDealt

        if (atk.force + dice > def.armorClass) {
            if (dice == 1) {
                damage = 0
            }
            if (dice == 20) {
                damage = atk.damageDealt * 2
            }
            return damage
        } else {
            return 0
        }
    }
}